### DISCUSSION TITLE
noob but legit question

### DISCUSSION THREAD

**Discussion by ripper6996** [kibook/pmms#35] [Q&A]
Hi.

First off I'm very new to coding stuff but I am very well versed in computer stuff so I can pick up things easily.

When I put in a certain site path in the allowed urls like the same way with ? and $ symbols it doesnt load that video

Config.allowedUrls = {
    "^https?://w?w?w?%.?youtube.com/.*$",
    "^https?://w?w?w?%.?youtu.be/.*$",
    "^https?://w?w?w?%.?twitch.tv/.*$",
}

Like this..

It doesnt load and gives an error like this, "DEMUXER ERROR COULD NOT FFmpegDemuxer: open context failed"

The site uses php I think.. I dont know if its a problem.. or the way I am putting the url in allowed urls is the problem.

As far as I know the site used HLS and it should work as per the description of your release :)

Anyway I could fix this? Pls do explain it like I am a total idiot or a noob.

**Comment by kibook** [comment]
Could you provide an example of a site you are trying to play media from? That is definitely the most important piece of information that is missing here.

Adding a URL template to `Config.allowedUrls` does not necessarily mean media from that site will work in pmms, this table exists solely to restrict which already compatible sites players can play media from. You may only want players to be able to play things from big sites like YouTube or Twitch because these sites are moderated and generally considered "safe". Allowing players to play media from any site has some risks associated with it.

pmms cannot play media from every possible website out there, as there are an infinite number of different ways in which sites can host media. pmms uses a library called [MediaElement.js](https://www.mediaelementjs.com) behind the scenes, so compatibility mainly comes down to whether that library supports the site in question.

pmms/MediaElement.js do support HLS streams, but it may be that the URL you are using is not the HLS stream itself, but a page containing a player which is displaying the HLS stream. Unless specific support is added to pmms/MediaElement.js for the site in question, you'd need to use the direct HLS stream URL instead.

For example, if you're using a URL like `https://somesite.com/videos/123`, perhaps the actual HLS stream playlist URLs would be something like `https://somesite.com/videos/123/streams/720p.m3u8` and `https://somesite.com/videos/123/streams/1080p.m3u8`, one for each quality you can select in the player on the site. Usually it's not quite that simple of course, but that's the general idea.

Most web browsers have a "Network" tab in their development tools which allows you to see all the resources a webpage is accessing, and this is often useful to find the direct link to a video or stream. If you can provide an example of a site you're trying to play media from, I can look further into what the site uses and whether or not it will be possible to use it with pmms.

**Comment by kowalskies** [comment]
Well. To be exact we arent trying to go to a very "safe" site and our server is 18+ only. We tried all the different sites that are the usual when it comes to 18+ videos(or p$rn). Even "The Hub" doesn't work. Anyway we can play videos from those sites?

I know this is gonna sound funny as hell but its really a feature we want for our server as we are an NSFW server and we want to push ourselves to be different from others.

As far as I can tell the "hub" uses php and HLS, but like you said but the content may be inaccessible to pmms.

Let me know if you can help. Thanks in advance.

**Comment by kibook** [comment]
I don't think MediaElement.js has support for PornHub or similar sites. While I have no issue with how you want to use pmms, I'm also not going to be spending any of my time trying to add support for porn sites.

There are some alternatives. You can try and get the direct video/HLS stream link from the site in question. One way to do this is to use the "Network" tab of your web browser's development tools. If the site uses HLS, there will typically be a `.m3u8` playlist being requested that lists all the different fragments of the video in order. You should be able to play this URL in pmms, although there is one caveat, and that is that pmms currently does not sync HLS streams between players, because it assumes they are live streams. Proper support for non-live HLS streams still needs to be added.

Another alternative would be to download and host certain videos on your own. You can place media in the `http/media` folder inside the resource, and then any files can be accessed with a relative URL. For example, if you place `pron.mp4` in that folder, you can play it in pmms by simply entering `pron.mp4` as the URL. You can also use your own external web server to host media, which is more recommended, as software like Apache and Nginx are much more robust and mature HTTP servers than what pmms uses to host videos directly on the FXServer, [httpmanager](https://github.com/kibook/httpmanager).

If you or someone else creates a MediaElement.js renderer for PornHub, or otherwise wants to add support for it in pmms, I'll gladly accept any pull requests.

**Reply by ripper6996** [reply]
So we tried putting files inside the `/pmms/http/media` folder and tried to play the file by putting the name of the file in the player, also tried with the path accurately with the `ip:port/pmms/http/media/file.mp4` as well but it doesnt play for some reason. Are we doing something wrong? The video is accessible when we try to play it in Chrome with the url to the servers ip but it doesnt play in pmms.

**Reply by itismejoey** [reply]
Did you ever figure this one out? Would like to be able to stream actual files, from the server, as well.

**Reply by kibook** [reply]
Were you using just the name of the file, without any of the directory names? E.g., if it's called `file.mp4`, that's all that should be entered into the URL field in the pmms UI.

For the other path you tried, did you include the protocol (`http://`) in the URL field? You must include that when specifying a full URL, otherwise pmms will interpret it as a relative URL inside of the `media` folder.

It could also have something to do with the format of the video. MPEG-4 is a container and can contain different types of video/audio, some of which may not be compatible with Chromium inside FiveM for whatever reason. Do you know what formats `file.mp4` is using?
